openedx/core/djangoapps/user_api/accounts/api.py

```diff
@@ -471,62 +782,87 @@ def _validate_password(password, username):
+    _validate_password_satisfies_max_length(password)
+
+
+def _validate_password_satisfies_min_length(password):
+    """Validate that the password is greater than or equal to
+    the minimum password length threshold.
+
+    :param password: The proposed password (unicode).
+    :return: None
+    :raises: AccountPasswordInvalid
+    """
     if len(password) < PASSWORD_MIN_LENGTH:
         raise AccountPasswordInvalid(
             u"Password must be at least {min} characters long".format(
                 min=PASSWORD_MIN_LENGTH
             )
         )
 
+
+def _validate_password_satisfies_max_length(password):
+    """Validate that the password is less than or equal to
+    the maximum password length threshold.
+
+    :param password: The proposed password (unicode).
+    :return: None
+    :raises: AccountPasswordInvalid
+    """
     if len(password) > PASSWORD_MAX_LENGTH:
         raise AccountPasswordInvalid(
             u"Password must be at most {max} characters long".format(
                 max=PASSWORD_MAX_LENGTH
             )
         )
 
-    if password == username:
-        raise AccountPasswordInvalid(u"Password cannot be the same as the username")
-
 
-def _validate_email(email):
-    """Validate the format of the email address.
+def _validate_password_works_with_username(password, username=None):
+    """Run validation checks on whether the password and username
+    go well together.
 
-    Arguments:
-        email (unicode): The proposed email.
-
-    Returns:
-        None
-
-    Raises:
-        AccountEmailInvalid
+    An example check is to see whether they are the same.
 
+    :param password: The proposed password (unicode).
+    :param username: The username associated with the user's account (unicode).
+    :return: None
+    :raises: AccountPasswordInvalid
     """
-    if not isinstance(email, basestring):
-        raise AccountEmailInvalid(u"Email must be a string")
-
-    if len(email) < EMAIL_MIN_LENGTH:
-        raise AccountEmailInvalid(
-            u"Email '{email}' must be at least {min} characters long".format(
-                email=email,
-                min=EMAIL_MIN_LENGTH
-            )
-        )
+    if password == username:
+        raise AccountPasswordInvalid(u"Password cannot be the same as the username")
 
-    if len(email) > EMAIL_MAX_LENGTH:
-        raise AccountEmailInvalid(
-            u"Email '{email}' must be at most {max} characters long".format(
-                email=email,
-                max=EMAIL_MAX_LENGTH
-            )
-        )
 
+def _validate_unicode(data):
     try:
-        validate_email(email)
-    except ValidationError:
-        raise AccountEmailInvalid(
-            u"Email '{email}' format is not valid".format(email=email)
-        )
+        if not isinstance(data, str) and not isinstance(data, unicode):
+            raise UnicodeError
+        # In some cases we pass the above, but it's still inappropriate utf-8.
+        str(data)
```

Comment:
What are we trying to accomplish here? I see what it does, but I don't understand _why_.